HIP-0013.md

```diff
@@ -0,0 +1,130 @@
+</code></pre>
+
+Records that contain spaces must be single quoted. Single quotes in the record can be escaped with a backslash:
+<pre><code><b>IDNS1</b> profile:location=<ins>'Berkeley, CA'</ins> profile:bio=<ins>'I\'m a passionate advocate for domain innovation.'</ins></code></pre>
+
+### Profile Records (`profile`)
+Human-readable user profile text data. 
+  - **`<key>`** The profile record name, any text key is valid, no spaces.
+  - **`<value>`** The profile record data.
+<pre><code><b>IDNS1 profile:&lt;key&gt;=&lt;value&gt;</b>
+
+<i>Examples:</i>
+IDNS1 profile:name=aox
+IDNS1 profile:avatar=https://hns.id/domain/aox.hns/avatar.png
+IDNS1 profile:location='Berkeley, CA'
+IDNS1 profile:email=hey@me.com
+</code></pre>
+
+### Service Records (`service`)
+Service keys must be made up of a reverse dot notation for a namespace which the service owns, for example, DNS names (e.g. .com, .io) or HNS name (i.e. .hns). Service Keys must contain at least one dot. See [EIP-634](https://eips.ethereum.org/EIPS/eip-634#service-keys).
+  - **`<key>`** A service key associated with the user, e.g. `com.github`.
+  - **`<identifier>`** Identifier for the service, e.g. Github username. 
+<pre><code><b>IDNS1 service:&lt;key&gt;=&lt;identifier&gt;</b>
+
+<i>Examples:</i>
+IDNS1 service:com.twitter=aoxborrow
+IDNS1 service:com.github=0xstefan
+IDNS1 service:io.keybase=pinheadmz
+</code></pre>
+
+### Wallet Records (`wallet`)
+Cryptocurrency wallet addresses. Uses [SLIP44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) to specify different blockchains.
+  - **`<cointype>`** The type of cryptocurrency, using SLIP44 cointype symbol or number, e.g. `hns` or `5353`
+  - **`<address>`** Wallet address for the specific cryptocurrency.
+<pre><code><b>IDNS1 wallet:&lt;cointype&gt;=&lt;address&gt;</b>
+
+<i>Examples:</i>
+IDNS1 wallet:btc=1Dvf5aQkVrVJeasRUDiLATGWLBGFWA9ovY
+IDNS1 wallet:hns=hs1qshuyulxra3pqpwr40303t8pn79232zztuk4qgz
+IDNS1 wallet:5353=hs1qshuyulxra3pqpwr40303t8pn79232zztuk4qgz
+IDNS1 wallet:eth=0xa08abA528da538cd3547a73C93d51B90201294B6
+</code></pre>
+
+
+### Contenthash Records (`contenthash`)
+A content hash or address for distributed networks like IPFS. See [ENSIP-7](https://docs.ens.domains/ensip/7).
+  - **`<protocol>`** Distributed network protocol name, e.g. `ipfs`.
+  - **`<address>`** Content address or content hash for distributed network.
+<pre><code><b>IDNS1 contenthash:&lt;protocol&gt;=&lt;address&gt;</b>
+
+<i>Examples:</i>
+IDNS1 contenthash:ipfs=c04dfad360be65ba134aab7d07ec59f0547a73ef3168
+IDNS1 contenthash:swarm=d191eb26786769f580809256b4685ef316805265ea1
+</code></pre>
+
+
+### Auth Records (`auth`)
+Authentication provider for use with "Login with Domain." See [Handshake Login](https://learn.namebase.io/development/guides/handshake-login/using-handshake-login#set-up-a-custom-identity-manager). (WIP)
+  - **`<provider>`** Identity provider URL.
+  - **`<fingerprint>`** Provider-specific fingerprint.
```

Comment:
I think having flexibility on hash algos, and future proofing (quantum) is needed as well (not a nice to have).

Reply:
@rithvikvibhu @jeefave is our guy who did the original NB OIDC. would be very open to contribution from either of you. what is the purpose for using the device/prefix as the subdomain host, and could it be moved to part of the TXT if there is room? e.g.
`HDNS1 auth:device={prefix} auth:fingerprint={hash} auth:keyid={keyid} auth:pubkey={pubkey}`

Reply:
iirc the idea behind the prefix was that since DNS doesn't* give you ALL records at once, adding a long enough pseudo-random prefix would mean nobody could enumerate find a list of all your devices. So if your laptop and phone stores keys, nobody can find the 2 records except the devices themselves since they store the device id locally in the browser.

When putting it on chain, it doesn't make a lot of sense since all records are already visible. So we do lose some privacy. Need to think about the format a bit more and honestly I don't really remember much about the protocol. btw hey John, been a while!

Reply:
the initial NB OIDC protocol definition contains a prevision to customize the hash algos ([here](https://learn.namebase.io/development/guides/handshake-login/handshake-based-oidc-authentication-protocol#security-considerations)). In the context of OAuth, the underlaying OIDC server will need to cater for the desired algorithms (currently it is [hardcoded](https://github.com/namebasehq/handshake-oidc/blob/7ae091ce697fdcfcee8c6ef2f4693aa3d8bfe900/src/hns.ts#L115)). It could be useful to explore how this could be extended to allow [signing with metamask ](https://docs.metamask.io/wallet/how-to/sign-data/) or others. 

The TXT format was inspired by these ietf drafts ([1](https://datatracker.ietf.org/doc/html/draft-sanz-openid-dns-discovery-01#section-3) and [2](https://datatracker.ietf.org/doc/html/draft-bertola-dns-openid-pidi-architecture-01#section-4.1)) where tag=value are separated by ";" and where values can be either text or base64 encoded strings depending on the nature of the tag.

I can confirm that the {prefix} subdomain aims to keep the list of the devices private because at the moment each devices generate their own private key. It seems like you remember quite well @rithvikvibhu :). Some similar mechanism could be leveraged to hide the list of services and profile infos if there are any privacy or security concerns? Another consideration is that using subdomains provides an elegant way to partition the data. It would mandate using DNSSEC. What are the benefits to store these data on chain?

What are your thoughts regarding this format?
```
<typehash>._hdns.<name> TXT <key>=<value>
servicehash1._hdns.name TXT v=0;identifier=aoxborrow
servicehash2._hdns.name TXT v=0;identifier=aoxborrow;picture=aHR0cHM6Ly9lbmNyeXB0ZWQtdGJuMC5nc3RhdGljLmNvbS9pbWFnZXM/cT10Ym46QU5kOUdjU1lzY2ZVQlVicXdHZF9ESFZoRy1aakNPRDdNVXB4cDR1aE5lN3RvVWc0dWcmcw==
```

Currently the HIP allows to define multiple provider fingerprints but the handshake login protocol sets the private key provider (idmanager) as a 1-1 association with its identity/domain. Relying parties are unaware of the signing party so identities can delegate to their preferred implementation. What use cases would require multiple id managers and how would they be resolved by the authication party?

![image](https://github.com/handshake-org/HIPs/assets/5779391/63f51239-a7c6-4f10-8e13-975e10056ae3)